### PR TITLE
Login and sign up auth

### PR DESCRIPTION
**Pull request by petarLits**
Finished for today. Login and sign up auth with bloc are done.

### CODE REVIEW

File: lib/data/firebase_auth_manager.dart [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import 'package:book/app_user.dart';
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:firebase_auth/firebase_auth.dart';
+
+class FirebaseAuthManager {
```
Comment (lecicdjuro) [inline lib/data/firebase_auth_manager.dart:5]: make Singleton class

File: lib/future_utils.dart [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+
+class FutureUtils {
```
Comment (lecicdjuro) [inline lib/future_utils.dart:4]: format

File: lib/login/bloc/login_bloc.dart [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import 'dart:async';
+
+import 'package:book/data/firebase_auth_manager.dart';
+import 'package:book/login/bloc/login_event.dart';
+import 'package:book/login/bloc/login_state.dart';
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+class LoginBloc extends Bloc<LoginEvent, LoginState> {
```
Comment (lecicdjuro) [inline lib/login/bloc/login_bloc.dart:8]: empty line

File: lib/future_utils.dart [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  static Future<dynamic> executeFutureWithLoader(
+      BuildContext context, Future<dynamic> task) async {
```
Comment (lecicdjuro) [inline lib/future_utils.dart:7]: add empty line

File: lib/login/bloc/login_event.dart [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  SignUpEvent({required this.user});
+
+  final AppUser user;
+
+  @override
+  List<Object?> get props => [user];
+}
+
+class Login extends LoginEvent {
+  Login({required this.email, required this.password});
```
Comment (lecicdjuro) [inline lib/login/bloc/login_event.dart]: comma at the end of last param

File: lib/login/bloc/login_state.dart [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+import 'package:equatable/equatable.dart';
+
+sealed class LoginState extends Equatable {
+  @override
+  List<Object?> get props => [];
+}
+
+class InitialState extends LoginState {
+  InitialState();
```
Comment (lecicdjuro) [inline lib/login/bloc/login_state.dart:11]: do we really need this?

File: lib/login/bloc/login_state.dart [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+class SuccessfulSignUp extends LoginState {
+  SuccessfulSignUp();
```
Comment (lecicdjuro) [inline lib/login/bloc/login_state.dart:15]: ditto?

File: lib/login/bloc/login_state.dart [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  
+  @override
+  List<Object?> get props => [Random().nextInt(10000)];
+}
+
+class SuccessfulLogin extends LoginState{
+  SuccessfulLogin();
+
+  @override
+  List<Object?> get props => [Random().nextInt(10000)];
```
Comment (lecicdjuro) [inline lib/login/bloc/login_state.dart:52]: why you put this value in props? Do we really need this?

File: lib/login/bloc/login_state.dart [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+class ErrorAuthState extends LoginState {
+  ErrorAuthState({required this.error});
+
+  final Exception error;
+  
+  @override
+  List<Object?> get props => [Random().nextInt(10000)];
```
Comment (lecicdjuro) [inline lib/login/bloc/login_state.dart]: instaed of random put error in props

File: lib/login/view/login_page.dart [hunk 10]
```diff
@@ -14,18 +19,49 @@ class _LoginPageState extends State<LoginPage> {
...
+              automaticallyImplyLeading: false,
+              backgroundColor: AppColors.primaryColor,
+              title: Text(AppLocalizations.of(context)!.loginTitle),
+            ),
+            body: _buildBody(),
+          );
+        },
+        listener: (context, state) {
+          if (state is SuccessfulLogin) {
+            Navigator.pushReplacement(
```
Comment (lecicdjuro) [inline lib/login/view/login_page.dart:47]: we should use named route

File: lib/login/view/login_page.dart [hunk 11]
```diff
@@ -14,18 +19,49 @@ class _LoginPageState extends State<LoginPage> {
...
+              context,
+              MaterialPageRoute(
+                builder: (context) => HomePage(),
+              ),
+            );
+          } else if (state is ErrorAuthState) {
+            final snackBar = SnackBar(
```
Comment (lecicdjuro) [inline lib/login/view/login_page.dart:54]: what is the color of Snackbar?

File: lib/login/view/login_page.dart [hunk 12]
```diff
@@ -140,7 +175,9 @@ class _LoginPageState extends State<LoginPage> {
                         Navigator.push(
                           context,
                           MaterialPageRoute(
-                            builder: (context) => SignUp(),
+                            builder: (context) => SignUp(
```
Comment (lecicdjuro) [inline lib/login/view/login_page.dart:180]: named route

File: lib/login/view/sign_up_page.dart [hunk 13]
```diff
@@ -1,8 +1,17 @@
...
+import 'package:book/app_user.dart';
+import 'package:book/login/bloc/login_bloc.dart';
+import 'package:book/login/bloc/login_event.dart';
+import 'package:book/login/bloc/login_state.dart';
 import 'package:flutter/material.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
 import 'package:flutter_gen/gen_l10n/app_localizations.dart';
 
 class SignUp extends StatefulWidget {
+  SignUp({required this.bloc});
```
Comment (lecicdjuro) [inline lib/login/view/sign_up_page.dart]: const constuctor

**Review by lecicdjuro — CHANGES_REQUESTED** [review]

File: lib/login/view/login_page.dart (hunk 10 above)
Comment (petarLits) [reply lib/login/view/login_page.dart:47]: named routes will be added in another pull request

File: lib/login/bloc/login_bloc.dart [hunk 14]
```diff
@@ -1,32 +1,34 @@
...
 import 'package:flutter_bloc/flutter_bloc.dart';
 
 class LoginBloc extends Bloc<LoginEvent, LoginState> {
+
   LoginBloc() : super(InitialState()) {
     on<SignUpEvent>(_onSignUpEvent);
     on<Login>(_onLogin);
   }
 
+  FirebaseAuthManager authManager = FirebaseAuthManager();
```
Comment (lecicdjuro) [inline lib/login/bloc/login_bloc.dart]: removi this

File: lib/login/bloc/login_bloc.dart [hunk 15]
```diff
@@ -1,32 +1,34 @@
...
+
   Future<void> _onSignUpEvent(
       SignUpEvent event, Emitter<LoginState> emit) async {
     try {
-      await FirebaseAuthManager.registerUser(event.user);
+      await authManager.registerUser(event.user);
```
Comment (lecicdjuro) [inline lib/login/bloc/login_bloc.dart]: instead of authManager=>FirebaseAuthManager()

File: lib/login/bloc/login_bloc.dart [hunk 16]
```diff
@@ -1,32 +1,34 @@
...
+      emit(SignUpErrorAuth(error: e));
     } on Exception catch (e) {
       emit(ErrorState(error: e));
     }
   }
 
   Future<void> _onLogin(Login event, Emitter<LoginState> emit) async {
     try {
-      await FirebaseAuthManager.loginUser(event.email, event.password);
+      await authManager.loginUser(event.email, event.password);
```
Comment (lecicdjuro) [inline lib/login/bloc/login_bloc.dart]: ditto

File: lib/login/view/sign_up_page.dart [hunk 17]
```diff
@@ -45,17 +45,20 @@ class _SignUpState extends State<SignUp> {
           if (state is SuccessfulSignUp) {
             Navigator.pop(context);
             final snackBar = SnackBar(
+              backgroundColor: AppColors.snackBar,
```
Comment (lecicdjuro) [inline lib/login/view/sign_up_page.dart]: better name for constant

**Review by lecicdjuro — CHANGES_REQUESTED** [review]

**Review by lecicdjuro — APPROVED** [review]

**Pull request merged**